rule/merge.go

```diff
@@ -138,6 +138,43 @@ func mergeAttrValues(srcAttr, dstAttr *attrValue) (bzl.Expr, error) {
 	return makePlatformStringsExpr(mergedExprs), nil
 }
 
+func shouldKeepRecursively(e bzl.Expr) bool {
+	if ShouldKeep(e) {
+		return true
+	}
+	switch e := e.(type) {
+	case *bzl.AssignExpr:
+		if shouldKeepRecursively(e.RHS) {
+			return true
+		}
+	case *bzl.DictExpr:
+		for _, kv := range e.List {
+			if shouldKeepRecursively(kv.Value) {
+				return true
+			}
+		}
+	case *bzl.ListExpr:
+		for _, el := range e.List {
+			if shouldKeepRecursively(el) {
+				return true
+			}
+		}
+	case *bzl.CallExpr:
+		for _, arg := range e.List {
+			if shouldKeepRecursively(arg) {
+				return true
+			}
+		}
+	case *bzl.BinaryExpr:
+		if shouldKeepRecursively(e.X) || shouldKeepRecursively(e.Y) {
+			return true
+		}
+	default:
+		fmt.Printf("unexpected expression type %T\n", e)
```

Comment:
Since we walk expressions recursively, wouldn't we always hit this eventually if there is no keep comment? Can we avoid recursing so deeply and e.g. only do it once?

Reply:
oops, this was for debugging. Let me get rid of it.

Reply:
Makes sense, thanks for the explanation!